File: cpp/309_BestTimetoBuyandSellStockwithCooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if(prices.empty()) return 0;
        int buy = -prices[0];
        int sell = 0,pre_sell = 0;
        for(int i = 1; i < prices.size(); ++i){
            int pre_buy = buy;
            buy = std::max(pre_sell - prices[i], pre_buy);
            pre_sell = sell;
            sell = std::max(pre_buy + prices[i], pre_sell);
        }
        return sell;     
    }
};

File: js/309_BestTimetoBuyandSellStockwithCooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    if(prices.length < 1) return 0;
    var buy = -prices[0];
    var sell = 0, preSell = 0;
    for(var i = 1; i < prices.length; ++i){
        var preBuy = buy;
        buy = Math.max(preSell - prices[i], preBuy);
        preSell = sell;
        sell = Math.max(preBuy+ prices[i], preSell);
    }
    return sell;
};